File: python/bubble_sort.py
#!/usr/bin/env python3
import sys
import time
from pathlib import Path


def bubble_sort(values: list[int]) -> None:
    n = len(values)
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            if values[j] > values[j + 1]:
                values[j], values[j + 1] = values[j + 1], values[j]
                swapped = True
        if not swapped:
            break


def checksum(values: list[int]) -> int:
    return sum(values)


def read_numbers(path: Path) -> list[int]:
    return [int(line.strip()) for line in path.read_text(encoding="utf-8").splitlines() if line.strip()]


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: bubble_sort.py <numbers_file>", file=sys.stderr)
        return 1

    values = read_numbers(Path(sys.argv[1]))
    start = time.perf_counter()
    bubble_sort(values)
    elapsed_ms = (time.perf_counter() - start) * 1000

    print(f"LANG=python N={len(values)} ELAPSED_MS={elapsed_ms:.3f} CHECKSUM={checksum(values)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: go/bubble_sort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func bubbleSort(values []int) {
	n := len(values)
	for i := 0; i < n; i++ {
		swapped := false
		for j := 0; j < n-i-1; j++ {
			if values[j] > values[j+1] {
				values[j], values[j+1] = values[j+1], values[j]
				swapped = true
			}
		}
		if !swapped {
			break
		}
	}
}

func checksum(values []int) int64 {
	var sum int64
	for _, v := range values {
		sum += int64(v)
	}
	return sum
}

func readNumbers(path string) ([]int, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var values []int
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		v, err := strconv.Atoi(line)
		if err != nil {
			return nil, err
		}
		values = append(values, v)
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return values, nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage: bubble_sort <numbers_file>")
		os.Exit(1)
	}

	values, err := readNumbers(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "read failed: %v\n", err)
		os.Exit(1)
	}

	start := time.Now()
	bubbleSort(values)
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1_000_000.0

	fmt.Printf("LANG=go N=%d ELAPSED_MS=%.3f CHECKSUM=%d\n", len(values), elapsedMs, checksum(values))
}
